pkg/web/middleware/csrf.go

```diff
@@ -0,0 +1,53 @@
+// Copyright © 2020 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package middleware
+
+import (
+	"github.com/gorilla/csrf"
+	echo "github.com/labstack/echo/v4"
+	"go.thethings.network/lorawan-stack/v3/pkg/webmiddleware"
+	"go.thethings.network/lorawan-stack/v3/pkg/webui"
+)
+
+// CSRF adds cross-site request forgery protection to the handler. The sync token
+// is passed to the template to be picked up by JavaScript.
+func CSRF(lookupName, path string, authKey []byte) echo.MiddlewareFunc {
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		pass := func(next echo.HandlerFunc) echo.HandlerFunc {
+			return func(c echo.Context) error {
+				token := csrf.Token(c.Request())
+				templateData := c.Get("template_data").(webui.TemplateData)
+				templateData.CSRFToken = token
+				c.Set("template_data", templateData)
+				c.Response().Header().Set("X-CSRF-Token", token)
+				return next(c)
+			}
+		}
+		return func(c echo.Context) error {
+			return echo.WrapMiddleware(
+				webmiddleware.CSRF(
+					nil,
+					authKey,
+					csrf.CookieName(lookupName),
+					csrf.FieldName(lookupName),
+					csrf.Secure(c.Request().TLS != nil),
```

Comment:
This may give incorrect behavior if we're behind a TLS-terminating reverse proxy. Our `RequestURL` middleware (`pkg/webmiddleware/request_url.go`) and `ProxyHeaders` middleware (`pkg/webmiddleware/proxy_headers.go`) set `(*http.Request).URL.Scheme` to `"https"` for this purpose.

Reply:
Hmm. If the Cookie is set to secure by default, it will not be included in non-TLS requests (meaning it would break the CSRF). Any idea what could be done about that, other than not using secure cookies?

Reply:
My suggestion was to replace `c.Request().TLS != nil` with `c.Request().URL.Scheme == "https"` because that's what we set for (forwarded) HTTPS requests.